File: tests/testCalcEngine.cpp
#include "core/calcEngine.hpp"
#include <cassert>
#include <cmath>
#include <iostream>
#include <vector>

static int passed = 0;
static int failed = 0;

#define ASSERT_NEAR(a, b, eps) \
    do { \
        if (std::fabs((a) - (b)) > (eps)) { \
            std::cerr << "FAIL: " << #a << " = " << (a) \
                      << ", expected " << (b) << " (line " << __LINE__ << ")\n"; \
            ++failed; \
        } else { ++passed; } \
    } while(0)

#define ASSERT_EQ(a, b) \
    do { \
        if ((a) != (b)) { \
            std::cerr << "FAIL: " << #a << " = " << (a) \
                      << ", expected " << (b) << " (line " << __LINE__ << ")\n"; \
            ++failed; \
        } else { ++passed; } \
    } while(0)

using namespace stockcalc;

void testBasicProfit() {
    PanelInput in{.totalInvestment = 1000.0, .sharePrice = 50.0,
                  .totalShares = 20.0, .targetPrice = 75.0};
    auto r = calculatePanel(in);
    ASSERT_NEAR(r.profit, 500.0, 0.01);
    ASSERT_NEAR(r.profitPlusInvest, 1500.0, 0.01);
    ASSERT_NEAR(r.gainPercent, 50.0, 0.01);
}

void testInferShares() {
    PanelInput in{.totalInvestment = 1000.0, .sharePrice = 25.0,
                  .lastChanged = 2, .secondLastChanged = 1};
    auto r = calculatePanel(in);
    ASSERT_NEAR(r.totalShares, 40.0, 0.01);
    ASSERT_EQ(r.inferredField, 3);
}

void testInferPrice() {
    PanelInput in{.totalInvestment = 500.0, .totalShares = 10.0,
                  .lastChanged = 3, .secondLastChanged = 1};
    auto r = calculatePanel(in);
    ASSERT_NEAR(r.sharePrice, 50.0, 0.01);
    ASSERT_EQ(r.inferredField, 2);
}

void testInferInvestment() {
    PanelInput in{.sharePrice = 100.0, .totalShares = 5.0,
                  .lastChanged = 3};
    auto r = calculatePanel(in);
    ASSERT_NEAR(r.totalInvestment, 500.0, 0.01);
    ASSERT_EQ(r.inferredField, 1);
}

void testZeroTarget() {
    PanelInput in{.totalInvestment = 1000.0, .sharePrice = 50.0, .totalShares = 20.0};
    auto r = calculatePanel(in);
    ASSERT_NEAR(r.profit, 0.0, 0.01);
    ASSERT_NEAR(r.profitPlusInvest, 0.0, 0.01);
}

void testNegativeProfit() {
    PanelInput in{.totalInvestment = 1000.0, .sharePrice = 50.0,
                  .totalShares = 20.0, .targetPrice = 30.0};
    auto r = calculatePanel(in);
    ASSERT_NEAR(r.profit, -400.0, 0.01);
    ASSERT_NEAR(r.gainPercent, -40.0, 0.01);
}

void testCombinedStats() {
    std::vector<PanelResult> panels;

    PanelResult p1{.totalInvestment = 1000.0, .sharePrice = 50.0,
                   .totalShares = 20.0, .profit = 500.0};
    PanelResult p2{.totalInvestment = 2000.0, .sharePrice = 100.0,
                   .totalShares = 20.0, .profit = -200.0};
    panels.push_back(p1);
    panels.push_back(p2);

    auto c = calculateCombined(panels);
    ASSERT_EQ(c.validCount, 2);
    ASSERT_NEAR(c.totalInvestment, 3000.0, 0.01);
    ASSERT_NEAR(c.totalShares, 40.0, 0.01);
    ASSERT_NEAR(c.totalProfit, 300.0, 0.01);
    ASSERT_NEAR(c.avgProfit, 150.0, 0.01);
    ASSERT_NEAR(c.avgSharePrice, 75.0, 0.01);
}

void testCombinedEmpty() {
    std::vector<PanelResult> panels;
    auto c = calculateCombined(panels);
    ASSERT_EQ(c.validCount, 0);
    ASSERT_NEAR(c.totalInvestment, 0.0, 0.01);
}

void testCombinedSkipsInvalid() {
    std::vector<PanelResult> panels;
    PanelResult valid{.totalInvestment = 500.0, .sharePrice = 25.0,
                      .totalShares = 20.0, .profit = 100.0};
    PanelResult invalid{};
    panels.push_back(valid);
    panels.push_back(invalid);

    auto c = calculateCombined(panels);
    ASSERT_EQ(c.validCount, 1);
    ASSERT_NEAR(c.totalInvestment, 500.0, 0.01);
}

void testInferFallbackTwoNonzero() {
    PanelInput in{.totalInvestment = 0.0, .sharePrice = 40.0, .totalShares = 10.0};
    auto r = calculatePanel(in);
    ASSERT_NEAR(r.totalInvestment, 400.0, 0.01);
    ASSERT_EQ(r.inferredField, 1);
}

void testLargeNumbers() {
    PanelInput in{.totalInvestment = 1000000.0, .sharePrice = 500.0,
                  .totalShares = 2000.0, .targetPrice = 750.0};
    auto r = calculatePanel(in);
    ASSERT_NEAR(r.profit, 500000.0, 0.01);
    ASSERT_NEAR(r.gainPercent, 50.0, 0.01);
}

void testFieldTracking() {
    ASSERT_EQ(isValidTrackingField(FieldId::totalInvestment), true);
    ASSERT_EQ(isValidTrackingField(FieldId::sharePrice), true);
    ASSERT_EQ(isValidTrackingField(FieldId::totalShares), true);
    ASSERT_EQ(isValidTrackingField(FieldId::targetPrice), false);
    ASSERT_EQ(isValidTrackingField(FieldId::none), false);
    ASSERT_EQ(isValidTrackingField(FieldId::profit), false);
    ASSERT_EQ(isValidTrackingField(FieldId::profitInvest), false);
}

int main() {
    testBasicProfit();
    testInferShares();
    testInferPrice();
    testInferInvestment();
    testZeroTarget();
    testNegativeProfit();
    testCombinedStats();
    testCombinedEmpty();
    testCombinedSkipsInvalid();
    testInferFallbackTwoNonzero();
    testLargeNumbers();
    testFieldTracking();

    std::cout << "\n=== Calc Engine Tests ===\n";
    std::cout << "Passed: " << passed << "\n";
    std::cout << "Failed: " << failed << "\n";
    std::cout << "Total:  " << (passed + failed) << "\n";

    return failed > 0 ? 1 : 0;
}

File: py_app/tests/test_calc_engine.py
"""Unit tests for the CalcEngine (Python fallback path)."""

import pytest
from py_app.core.bridge import CalcEngine, PanelResult


@pytest.fixture
def engine():
    e = CalcEngine()
    return e


class TestSmartInference:
    def test_infer_shares_from_investment_and_price(self, engine):
        r = engine.calculate_panel(10000, 200, 0, 250)
        assert r.total_shares == pytest.approx(50.0)
        assert r.inferred_field == 3

    def test_infer_price_from_investment_and_shares(self, engine):
        r = engine.calculate_panel(10000, 0, 50, 300)
        assert r.share_price == pytest.approx(200.0)
        assert r.inferred_field == 2

    def test_infer_investment_from_price_and_shares(self, engine):
        r = engine.calculate_panel(0, 200, 50, 0)
        assert r.total_investment == pytest.approx(10000.0)
        assert r.inferred_field == 1

    def test_no_inference_all_zeros(self, engine):
        r = engine.calculate_panel(0, 0, 0, 0)
        assert r.inferred_field == 0

    def test_no_inference_single_field(self, engine):
        r = engine.calculate_panel(10000, 0, 0, 0)
        assert r.inferred_field == 0


class TestLastChangedTracking:
    def test_last_investment_derives_shares(self, engine):
        r = engine.calculate_panel(10000, 200, 100, 0, last_changed=1)
        assert r.total_shares == pytest.approx(50.0)
        assert r.inferred_field == 3

    def test_last_shares_second_investment_derives_price(self, engine):
        r = engine.calculate_panel(10000, 200, 40, 0, last_changed=3, second_last_changed=1)
        assert r.share_price == pytest.approx(250.0)
        assert r.inferred_field == 2

    def test_last_price_derives_shares(self, engine):
        r = engine.calculate_panel(10000, 250, 50, 0, last_changed=2)
        assert r.total_shares == pytest.approx(40.0)
        assert r.inferred_field == 3

    def test_last_shares_no_second_derives_investment(self, engine):
        r = engine.calculate_panel(0, 200, 50, 0, last_changed=3)
        assert r.total_investment == pytest.approx(10000.0)
        assert r.inferred_field == 1


class TestProfitCalculation:
    def test_positive_profit(self, engine):
        r = engine.calculate_panel(10000, 200, 0, 250)
        assert r.profit == pytest.approx(2500.0)
        assert r.profit_plus_invest == pytest.approx(12500.0)
        assert r.gain_percent == pytest.approx(25.0)

    def test_negative_profit(self, engine):
        r = engine.calculate_panel(10000, 200, 0, 150)
        assert r.profit == pytest.approx(-2500.0)
        assert r.gain_percent == pytest.approx(-25.0)

    def test_zero_target_no_profit(self, engine):
        r = engine.calculate_panel(10000, 200, 0, 0)
        assert r.profit == pytest.approx(0.0)
        assert r.profit_plus_invest == pytest.approx(0.0)

    def test_breakeven(self, engine):
        r = engine.calculate_panel(10000, 200, 0, 200)
        assert r.profit == pytest.approx(0.0)
        assert r.gain_percent == pytest.approx(0.0)


class TestCombinedCalculation:
    def test_two_valid_panels(self, engine):
        panels = [
            PanelResult(10000, 200, 50, 12500, 2500, 25, 0),
            PanelResult(5000, 100, 50, 7500, 2500, 50, 0),
        ]
        c = engine.calculate_combined(panels)
        assert c.valid_count == 2
        assert c.total_investment == pytest.approx(15000.0)
        assert c.total_shares == pytest.approx(100.0)
        assert c.total_profit == pytest.approx(5000.0)
        assert c.avg_profit == pytest.approx(2500.0)
        assert c.avg_share_price == pytest.approx(150.0)

    def test_skips_invalid_panel(self, engine):
        panels = [
            PanelResult(10000, 200, 50, 0, 2500, 0, 0),
            PanelResult(0, 0, 0, 0, 0, 0, 0),
        ]
        c = engine.calculate_combined(panels)
        assert c.valid_count == 1
        assert c.total_investment == pytest.approx(10000.0)

    def test_empty_list(self, engine):
        c = engine.calculate_combined([])
        assert c.valid_count == 0
        assert c.total_investment == pytest.approx(0.0)

    def test_weighted_avg_price(self, engine):
        panels = [
            PanelResult(10000, 100, 100, 0, 0, 0, 0),
            PanelResult(10000, 200, 50, 0, 0, 0, 0),
        ]
        c = engine.calculate_combined(panels)
        expected_avg = (100 * 100 + 200 * 50) / 150
        assert c.avg_share_price == pytest.approx(expected_avg)
